src/celeritas/phys/AtomicNumber.hh

```diff
@@ -0,0 +1,85 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/phys/AtomicNumber.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Type-safe atomic number identifier.
+ *
+ * Atomic numbers have constraints that integers do not (zero is not an atomic
+ * number, dividing one atomic number by another is meaningless) so this type
+ * improves the semantics and safety of Z.
+ *
+ * Generally speaking, Z numbers are used during setup and when evaluating
+ * expressions that use atomic charge. Z should not be used to "index" data
+ * during runtime.
+ *
+ * Constructing with a nonpositive Z gives the result a "false" state (\c get
+ * will throw a \c DebugError).
+ */
+class AtomicNumber
+{
+  public:
+    //! Construct with an invalid/unassigned value of zero
+    constexpr AtomicNumber() = default;
+
+    //! Construct with the Z value
+    explicit CELER_CONSTEXPR_FUNCTION AtomicNumber(int z_value) : z_(z_value)
+    {
+    }
+
+    //! True if value is assigned/valid
+    explicit CELER_CONSTEXPR_FUNCTION operator bool() const { return z_ > 0; }
+
+    //! Get the Z value
+    CELER_CONSTEXPR_FUNCTION int unchecked_get() const { return z_; }
+
+    // Get the Z value
+    inline CELER_FUNCTION int get() const;
+
+  private:
+    int z_{0};
```

Comment:
Any particular reason to not make it unsigned?

Reply:
I don't think it makes a whit of difference, but generally `int` is for general integer values and `unsigned` is primarily for sizes. Since the range of valid values for `AtomicNumber` isn't the same as the range for either of those, it really shouldn't matter matter.

The only difference might be that when you multiply the `.get()` value into another expression the "range" of int is more close to the "range" of double (both allow signed values) or do a comparison to an unsigned int (which you shouldn't) as some compiler settings lead to warnings there.

The other thing about `int` is that by definition overflowing is undefined behavior, whereas with `unsigned` it's required to wrap around. This lets compilers apply certain optimizations sometimes... though I can't imagine they'd apply in this case.

Reply:
Makes sense. Interesting... I thought ints would wrap around like unsigned ints.

Reply:
In practice they usually do, but because it's UB they don't have to, and furthermore the compiler can make assumptions that they never will. http://kristerw.blogspot.com/2016/02/how-undefined-signed-overflow-enables.html